samples/subsys/instrumentation/src/main.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2023 Linaro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/instrumentation/instrumentation.h>
+
+/*
+ * This sample shows the instrumentation subsystem tracing and profiling
+ * features. It basically consists in two threads in a ping-pong mode, taking
+ * turns to execute loops that spend some CPU cycles.
+ *
+ * After one flashes this sample to the target, it must be possible to
+ * collect and visualize traces and profiling info using the instrumentation CLI
+ * tool, zaru. Please _note_ that this subsystem uses the retained_mem driver,
+ * hence it's necessary to add the proper devicetree overlay for the target
+ * board (see './boards/b_u585i_iot02a.overlay' in this dir. as a reference).
+ *
+ * Example:
+ *
+ * $ export ZEPHYR_BASE=~/zephyrproject/zephyr
+ *
+ * $ west build -p always -b b_u585i_iot02a samples/subsys/instrumentation -t flash
+ * (wait ~2 seconds so the sample finishes 2 rounds of ping-pong between 'main'
+ * and 'thread_A')
+ *
+ * (check instrumentation status)
+ * $ scripts/zaru status
+ *
+ * (set the tracing / profiling location; in this sample the function
+ * 'get_sem_and_exec_function' is the one interesting to allow the observation
+ * of context switches)
+ * $ scripts/zaru trace -v -c get_sem_and_exec_function
+ *
+ * (reboot target so tracing/profiling at the location is effective)
+ * $ scripts/zaru reboot
+ * (wait ~2 seconds so the sample finishes 2 rounds of ping-pong between 'main'
+ * and 'thread_A')
+ *
+ * (get traces)
+ * $ scripts/zaru trace -v
+ *
+ * (.. and get profile)
+ * $ scripts/zaru profile -v -n 10
+ *
+ * Or alternatively, export the traces to Perfetto:
+ *
+ * (it's necessary to reboot because 'zaru trace' dumped the buffer and it's now
+ * empty)
+ * $ scripts/zaru reboot
+ * (wait ~2 seconds so the sample finishes 2 rounds of ping-pong between 'main'
+ * and 'thread_A')
+ *
+ * $ scripts/zaru trace -v --perfetto --output perfetto_zephyr.json
+ * (then go to http://perfetto.dev, Trace Viewer, and load perfetto_zephyr.json)
+ *
+ */
+
+#define SLEEPTIME 10
+#define STACKSIZE 1024
+#define PRIORITY  7
+
+void __attribute__((optimize("O0"))) loop_0(void)
```

Comment:
We have optimize O0 attribute define which is suitable for all toolchains, so we can use it here:

```suggestion
void __no_optimization loop_0(void)
```

Reply:
@evgeniy-paltsev oh, cool! Thanks! :+1:

Reply:
Done.